### ISSUE TITLE
Cannot access to Heapster information with default dashboard yaml at kubeadm

### ISSUE THREAD

**Issue by s-miyoshi-fj** [kubernetes/dashboard#2224]
Hi all,

I have trouble about dashboard UI.
I cannot see Heapster information such as CPU usage and memory usage in dashboard.
But, I can see normal information such as pods status and deployments status.

I think this is not heapster problem, because I can see Heapster information by grafana.

I built a kubernetes environment by kubeadm.
Here is my dashboard yaml file.
It is almost same as default file, but I added NodePort setting in 'kind: Service'.

```kubernetes-dashboard.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    k8s-app: kubernetes-dashboard
  name: kubernetes-dashboard
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: kubernetes-dashboard
  labels:
    k8s-app: kubernetes-dashboard
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: kubernetes-dashboard
  namespace: kube-system
---
kind: Deployment
apiVersion: extensions/v1beta1
metadata:
  labels:
    k8s-app: kubernetes-dashboard
  name: kubernetes-dashboard
  namespace: kube-system
spec:
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: kubernetes-dashboard
  template:
    metadata:
      labels:
        k8s-app: kubernetes-dashboard
    spec:
      containers:
      - name: kubernetes-dashboard
        image: gcr.io/google_containers/kubernetes-dashboard-amd64:v1.6.3
        ports:
        - containerPort: 9090
          protocol: TCP
        args:
          # Uncomment the following line to manually specify Kubernetes API server Host
          # If not specified, Dashboard will attempt to auto discover the API server and connect
          # to it. Uncomment only if the default does not work.
          # - --apiserver-host=http://my-address:port
        livenessProbe:
          httpGet:
            path: /
            port: 9090
          initialDelaySeconds: 30
          timeoutSeconds: 30
      serviceAccountName: kubernetes-dashboard
      # Comment the following tolerations if Dashboard must not be deployed on master
      tolerations:
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
---
kind: Service
apiVersion: v1
metadata:
  labels:
    k8s-app: kubernetes-dashboard
  name: kubernetes-dashboard
  namespace: kube-system
spec:
  ports:
  - port: 80
    targetPort: 9090
    nodePort: 31000
  type: NodePort
  selector:
    k8s-app: kubernetes-dashboard
```

My environment information is follows.

\# cat /etc/redhat-release
```
CentOS Linux release 7.2.1511 (Core)
```
\# kubeadm version
```
kubeadm version: &version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-06-29T22:55:19Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
\# kubectl get all -n kube-system
```
NAME                                                  READY     STATUS    RESTARTS   AGE
po/heapster-1456382540-48p82                          1/1       Running   0          7d
po/kube-apiserver-k8s-master-lsoft-miyoshi            1/1       Running   0          8d
po/kube-controller-manager-k8s-master-lsoft-miyoshi   1/1       Running   1          8d
po/kube-dns-2425271678-075vz                          3/3       Running   15         8d
po/kube-flannel-ds-pcm2h                              2/2       Running   0          8d
po/kube-flannel-ds-tb5dl                              2/2       Running   8          8d
po/kube-proxy-wllk8                                   1/1       Running   4          8d
po/kube-proxy-xw0km                                   1/1       Running   0          8d
po/kube-scheduler-k8s-master-lsoft-miyoshi            1/1       Running   2          8d
po/kubernetes-dashboard-3313488171-xjhhc              1/1       Running   0          31m
po/monitoring-grafana-1545420726-8200g                1/1       Running   0          1d
po/monitoring-influxdb-1870447071-6m93x               1/1       Running   0          7d

NAME                       CLUSTER-IP       EXTERNAL-IP   PORT(S)         AGE
svc/heapster               10.96.230.120    <none>        80/TCP          7d
svc/kube-dns               10.96.0.10       <none>        53/UDP,53/TCP   8d
svc/kubernetes-dashboard   10.108.85.78     <nodes>       80:31000/TCP    31m
svc/monitoring-grafana     10.100.201.219   <nodes>       80:31001/TCP    1d
svc/monitoring-influxdb    10.109.100.230   <none>        8086/TCP        7d

NAME                          DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deploy/heapster               1         1         1            1           7d
deploy/kube-dns               1         1         1            1           8d
deploy/kubernetes-dashboard   1         1         1            1           31m
deploy/monitoring-grafana     1         1         1            1           1d
deploy/monitoring-influxdb    1         1         1            1           7d

NAME                                 DESIRED   CURRENT   READY     AGE
rs/heapster-1456382540               1         1         1         7d
rs/kube-dns-2425271678               1         1         1         8d
rs/kubernetes-dashboard-3313488171   1         1         1         31m
rs/monitoring-grafana-1545420726     1         1         1         1d
rs/monitoring-influxdb-1870447071    1         1         1         7d
```

Please tell me about what's I have to do next?

**Comment by ffahri** [comment]
you can try to add ths role controllecr yaml . That was solve my problem.
https://github.com/kubernetes/heapster/blob/master/deploy/kube-config/rbac/heapster-rbac.yaml

**Comment by floreks** [comment]
Please give us logs from dashboard container and from heapster `kubectl logs <pod_name>`

**Comment by s-miyoshi-fj** [comment]
I added following role controller yaml.
This is almost same as yaml of ffahri's link, but I changed 'roleRef.name'
But I still have the same problem.
(system:heapster -> cluster-admin)

```
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: heapster
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: heapster
  namespace: kube-system
```

Is 'roleRef.name' important for this problem?


And here is kubectl log.
1. heapster pod

```
I0802 08:03:35.450522       1 heapster.go:72] /heapster --source=kubernetes:https://kubernetes.default.svc.cluster.local --sink=influxdb:http://monitoring-influxdb.kube-system.svc.cluster.local:8086
I0802 08:03:35.450590       1 heapster.go:73] Heapster version v1.3.0
I0802 08:03:35.451128       1 configs.go:61] Using Kubernetes client with master "https://kubernetes.default.svc.cluster.local" and version v1
I0802 08:03:35.451145       1 configs.go:62] Using kubelet port 10255
I0802 08:03:35.603496       1 influxdb.go:252] created influxdb sink with options: host:monitoring-influxdb.kube-system.svc.cluster.local:8086 user:root db:k8s
I0802 08:03:35.603551       1 heapster.go:196] Starting with InfluxDB Sink
I0802 08:03:35.603558       1 heapster.go:196] Starting with Metric Sink
I0802 08:03:35.621964       1 heapster.go:106] Starting heapster on port 8082
I0802 08:04:05.251244       1 influxdb.go:215] Created database "k8s" on influxDB server at "monitoring-influxdb.kube-system.svc.cluster.local:8086"
```

2. dashboard pod

```
Using HTTP port: 8443
Using in-cluster config to connect to apiserver
Using service account token for csrf signing
No request provided. Skipping authorization header
Successful initial request to the apiserver, version: v1.7.2
No request provided. Skipping authorization header
Creating in-cluster Heapster client
Could not enable metric client: Health check failed: an error on the server ("<html><head>\r\n<title>Network Error</TITLE>\r\n</HEAD>\r\n<body>\r\n<font face=\"Helvetica\">\r\n<big><strong></strong></big><br>\r\n</FONT>\r\n<blockquote>\r\n<table border=\"0\" cellpadding=\"1\" width=\"80%\">\r\n<tr><td>\r\n<font face=\"Helvetica\">\r\n<big>Network Error (tcp_error)</big>\r\n<br>\r\n<br>\r\n</FONT>\r\n</TD></TR>\r\n<tr><td>\r\n<font face=\"Helvetica\">\r\nA communication error occurred: \"No route to host\"\r\n</FONT>\r\n</TD></TR>\r\n<tr><td>\r\n<font face=\"Helvetica\">\r\nThe Web Server may be down, too busy, or experiencing other problems preventing it from responding to requests. You may wish to try again at a later time.\r\n</FONT>\r\n</TD></TR>\r\n<tr><td>\r\n<font face=\"Helvetica\" size=\"2\">\r\n<br>\r\n\r\n</FONT>\r\n</TD></TR>\r\n</TABLE>\r\n</blockquote>\r\n</FONT>\r\n</BODY></HTML>") has prevented the request from succeeding (get services heapster). Continuing.
Getting application global configuration
Application configuration {"serverTime":1501661026481}
[2017-08-02T08:03:47Z] Incoming HTTP/1.1 GET /api/v1/thirdpartyresource request from 172.30.0.1:59132: {}
[2017-08-02T08:03:47Z] Incoming HTTP/1.1 GET /api/v1/rbac/status request from 172.30.0.1:59131: {}
Getting list of third party resources
[2017-08-02T08:03:47Z] Outcoming response to 172.30.0.1:59132 with 200 status code
[2017-08-02T08:03:47Z] Outcoming response to 172.30.0.1:59131 with 200 status code
[2017-08-02T08:03:47Z] Incoming HTTP/1.1 GET /api/v1/workload?filterBy=&itemsPerPage=15&page=1&sortBy=d,creationTimestamp request from 172.30.0.1:59131: {}
Getting lists of all workloads
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
Getting pod metrics
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
[restful] 2017/08/02 08:03:47 log.go:26: No metric client provided. Skipping metrics.
[2017-08-02T08:03:47Z] Outcoming response to 172.30.0.1:59131 with 200 status code

```

**Comment by floreks** [comment]
There is some problem with pod to pod communication in your cluster.
```
Could not enable metric client: Health check failed: an error on the server ("<html><head>\r\n<title>Network Error</TITLE>\r\n</HEAD>\r\n<body>\r\n<font face=\"Helvetica\">\r\n<big><strong></strong></big><br>\r\n</FONT>\r\n<blockquote>\r\n<table border=\"0\" cellpadding=\"1\" width=\"80%\">\r\n<tr><td>\r\n<font face=\"Helvetica\">\r\n<big>Network Error (tcp_error)</big>\r\n<br>\r\n<br>\r\n</FONT>\r\n</TD></TR>\r\n<tr><td>\r\n<font face=\"Helvetica\">\r\nA communication error occurred: \"No route to host\"\r\n</FONT>\r\n</TD></TR>\r\n<tr><td>\r\n<font face=\"Helvetica\">\r\nThe Web Server may be down, too busy, or experiencing other problems preventing it from responding to requests. You may wish to try again at a later time.\r\n</FONT>\r\n</TD></TR>\r\n<tr><td>\r\n<font face=\"Helvetica\" size=\"2\">\r\n<br>\r\n\r\n</FONT>\r\n</TD></TR>\r\n</TABLE>\r\n</blockquote>\r\n</FONT>\r\n</BODY></HTML>") has prevented the request from succeeding (get services heapster). Continuing.
```

Heapster is not responding to a health check.

**Comment by s-miyoshi-fj** [comment]
I solved this issue by oneself.
floreks is right, it was communication error to heapster.

I added dashboard argument '--heapster-host' in yaml.
This is my yaml of 'kind:Deployment'.

```
kind: Deployment
apiVersion: extensions/v1beta1
metadata:
  labels:
    k8s-app: kubernetes-dashboard
  name: kubernetes-dashboard
  namespace: kube-system
spec:
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: kubernetes-dashboard
  template:
    metadata:
      labels:
        k8s-app: kubernetes-dashboard
    spec:
      containers:
      - name: kubernetes-dashboard
        image: gcr.io/google_containers/kubernetes-dashboard-amd64:v1.6.3
        ports:
        - containerPort: 9090
          protocol: TCP
        args:
          # Uncomment the following line to manually specify Kubernetes API server Host
          # If not specified, Dashboard will attempt to auto discover the API server and connect
          # to it. Uncomment only if the default does not work.
          # - --apiserver-host=http://my-address:port
          - --heapster-host=http://heapster.kube-system.svc.cluster.local
        livenessProbe:
          httpGet:
            path: /
            port: 9090
          initialDelaySeconds: 30
          timeoutSeconds: 30
      serviceAccountName: kubernetes-dashboard
      # Comment the following tolerations if Dashboard must not be deployed on master
      #tolerations:
      #- key: node-role.kubernetes.io/master
      #  effect: NoSchedule
```

Thank you!

**Comment by kumudt** [comment]
I thought this issue was fixed with 1.6.3 update. It's still there I guess.

**Comment by floreks** [comment]
It is solved. This is a cluster issue. In 1.6.2 dashboard was building wrong path to heapster and there was always the same error that health check to heapster was failing because dashboard could not access it. Error message that @s-miyoshi-fj got is completely different and not related to our check.

**Comment by jamstar** [comment]
so i got the same issue and solved it the same way as the original submitter.  Im not using kubeadm, i actually use a custom cluster.  someone else had the same problem that was using kops.   Is it possible there is something else similar going on?

**Comment by floreks** [comment]
If in-cluster discovery is not working with Dashboard 1.6.3 then it means you have a networking issue in your cluster.

**Comment by jamstar** [comment]
so it seems to be a timing issue with dashboard.   im running a custom cluster and use vagrant with virtualbox for local development with ansible.     i deploy all my addons at the same time when i build the cluster. Dashboard doesn't have the metrics when they are all created at the same time with/without the dns argument.   If you wait a short bit and kill the pod to restart it, dashboard then has them.

**Comment by floreks** [comment]
Yes, this issue was fixed already and will be available in release 1.7. Our wiki pages describe mostly functionality of Dashboard 1.7.

**Comment by jamstar** [comment]
great.  thank you for the info.  1.7 has a bunch of interesting changes.
